### ISSUE TITLE
How to extract features in forward method.

### ISSUE THREAD

**Issue by junwoony** [skorch-dev/skorch#428]
I am using `NeuralNetRegressor` to predict output values and I am wondering If there is a way to extract/visualize intermediate features in the forward method (such as `crys_fea` in the forward). It seems like `NeuralNetRegressor` does not support multiple return values in the forward method. Please give me some suggestions to extract those intermediate features.

```python
class ConvNet(nn.Module):

    def __init__(self, ...)
        super(CrystalGraphConvNet, self).__init__()
        self.embedding = nn.Linear(orig_atom_fea_len, atom_fea_len)
        self.convs = nn.ModuleList([ConvLayer()])
        self.conv_to_fc = nn.Linear(atom_fea_len, h_fea_len)
        self.conv_to_fc_softplus = nn.Softplus()
        self.fc_out = nn.Linear(h_fea_len, 1)

    def forward(self, ...):
        atom_fea = self.embedding(atom_fea)
        for conv_func in self.convs:
            atom_fea = conv_func(atom_fea, nbr_fea, nbr_fea_idx)
        crys_fea = self.pooling(atom_fea, crystal_atom_idx)
        crys_fea = self.conv_to_fc(self.conv_to_fc_softplus(crys_fea))
        out = self.fc_out(crys_fea)
        return out
```

```python
from skorch import NeuralNetRegressor 
from model import CrystalGraphConvNet

net = NeuralNetRegressor(
    CrystalGraphConvNet,
    modules...)

net.initialize()
net.fit(SDT_training,target_training)
```

**Comment by ottonemo** [comment]
For a quick inspection you can also use a PyTorch [foward hook][1] which will be called whenever the `forward` method of a specific module is called:

```python
net.initialize()
net.module_.conv_to_fc.register_forward_hook(lambda mod, inp, out: print(out))
net.fit(SDT_training,target_training)
```

[1]: https://pytorch.org/docs/master/nn.html#torch.nn.Module.register_forward_hook

**Comment by junwoony** [comment]
Thanks for your inputs, @BenjaminBossan and @ottonemo.
I tried to set `return out, crys_fea` in the forward, but it gives me an error if I try to run `net.fit(SDT_training, traget_training)`.

The error says that,
```
TypeError                                 Traceback (most recent call last)
<ipython-input-11-368234cd854e> in <module>()
      1 net.initialize()
      2 # net.predict(SDT_training)
----> 3 net.fit(SDT_training,target_training)

/home/zulissi/miniconda3/envs/cgcnn/lib/python3.6/site-packages/skorch/regressor.py in fit(self, X, y, **fit_params)
     88         # this is actually a pylint bug:
     89         # https://github.com/PyCQA/pylint/issues/1085
---> 90         return super(NeuralNetRegressor, self).fit(X, y, **fit_params)

/home/zulissi/miniconda3/envs/cgcnn/lib/python3.6/site-packages/skorch/net.py in fit(self, X, y, **fit_params)
    840             self.initialize()
    841 
--> 842         self.partial_fit(X, y, **fit_params)
    843         return self
    844 

/home/zulissi/miniconda3/envs/cgcnn/lib/python3.6/site-packages/skorch/net.py in partial_fit(self, X, y, classes, **fit_params)
    799         self.notify('on_train_begin', X=X, y=y)
    800         try:
--> 801             self.fit_loop(X, y, **fit_params)
    802         except KeyboardInterrupt:
    803             pass

/home/zulissi/miniconda3/envs/cgcnn/lib/python3.6/site-packages/skorch/net.py in fit_loop(self, X, y, epochs, **fit_params)
    737                 yi_res = yi if not y_train_is_ph else None
    738                 self.notify('on_batch_begin', X=Xi, y=yi_res, training=True)
--> 739                 step = self.train_step(Xi, yi, **fit_params)
    740                 self.history.record_batch('train_loss', step['loss'].item())
    741                 self.history.record_batch('train_batch_size', get_len(Xi))

/home/zulissi/miniconda3/envs/cgcnn/lib/python3.6/site-packages/skorch/net.py in train_step(self, Xi, yi, **fit_params)
    666             step_accumulator.store_step(step)
    667             return step['loss']
--> 668         self.optimizer_.step(step_fn)
    669         return step_accumulator.get_step()
    670 

/home/zulissi/miniconda3/envs/cgcnn/lib/python3.6/site-packages/torch/optim/adam.py in step(self, closure)
     55         loss = None
     56         if closure is not None:
---> 57             loss = closure()
     58 
     59         for group in self.param_groups:

/home/zulissi/miniconda3/envs/cgcnn/lib/python3.6/site-packages/skorch/net.py in step_fn()
    663         step_accumulator = self.get_train_step_accumulator()
    664         def step_fn():
--> 665             step = self.train_step_single(Xi, yi, **fit_params)
    666             step_accumulator.store_step(step)
    667             return step['loss']

/home/zulissi/miniconda3/envs/cgcnn/lib/python3.6/site-packages/skorch/net.py in train_step_single(self, Xi, yi, **fit_params)
    606         self.optimizer_.zero_grad()
    607         y_pred = self.infer(Xi, **fit_params)
--> 608         loss = self.get_loss(y_pred, yi, X=Xi, training=True)
    609         loss.backward()
    610 

/home/zulissi/miniconda3/envs/cgcnn/lib/python3.6/site-packages/skorch/net.py in get_loss(self, y_pred, y_true, X, training)
   1069         """
   1070         y_true = to_tensor(y_true, device=self.device)
-> 1071         return self.criterion_(y_pred, y_true)
   1072 
   1073     def get_dataset(self, X, y=None):

/home/zulissi/miniconda3/envs/cgcnn/lib/python3.6/site-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    475             result = self._slow_forward(*input, **kwargs)
    476         else:
--> 477             result = self.forward(*input, **kwargs)
    478         for hook in self._forward_hooks.values():
    479             hook_result = hook(self, input, result)

/home/zulissi/miniconda3/envs/cgcnn/lib/python3.6/site-packages/torch/nn/modules/loss.py in forward(self, input, target)
    419 
    420     def forward(self, input, target):
--> 421         return F.mse_loss(input, target, reduction=self.reduction)
    422 
    423 

/home/zulissi/miniconda3/envs/cgcnn/lib/python3.6/site-packages/torch/nn/functional.py in mse_loss(input, target, size_average, reduce, reduction)
   1714     else:
   1715         reduction = _Reduction.get_enum(reduction)
-> 1716     return _pointwise_loss(lambda a, b: (a - b) ** 2, torch._C._nn.mse_loss, input, target, reduction)
   1717 
   1718 

/home/zulissi/miniconda3/envs/cgcnn/lib/python3.6/site-packages/torch/nn/functional.py in _pointwise_loss(lambd, lambd_optimized, input, target, reduction)
   1672         return torch.mean(d) if reduction == 'elementwise_mean' else torch.sum(d)
   1673     else:
-> 1674         return lambd_optimized(input, target, reduction)
   1675 
   1676 

TypeError: mse_loss(): argument 'input' (position 1) must be Tensor, not tuple
```

**Comment by junwoony** [comment]
@BenjaminBossan,
I'm sorry but I think I'm kind of stuck. If I fit the model using `.fit()` with 
```
    def forward(self, ...):
        ...
        self.features = crys_fea
        ...
        return out
```
since I only want the features after the last epoch.
Then how can I print out `self.features = crys_fea` after training is completed?

**Comment by BenjaminBossan** [comment]
You should be able to access it like this:

`net.module_.features`

**Comment by junwoony** [comment]
Thanks @BenjaminBossan.
It seems like it is working!
I just have one more question. What I did above is for the last batch.
What if I want the features from the entire batches at the last epoch?
Let's say my training has 20,000 training samples, and fractions of them will be assigned to multiple batches. And I want the features (`crys_fea`) for all of those 20,000 samples. Thens should I go with,
```
features = []

class ConvNet(nn.Module):
    def __init__(self,...):
        ...
    def forward(self, ...):
         ...
         features.append(crys_fea)  # <== new line
         out = self.fc_out(crys_fea)
         return out
```

If so, how can I print out the features list? 
Here, `net.module_.features` doesn't seem it works since it's not `self.features`.
I would really appreciate your help.

**Comment by BenjaminBossan** [comment]
> What if I want the features from the entire batches at the last epoch?

I guess for that, another approach is easier. Subclass the `NeuralNetRegressor` and return both outputs (i.e. the first suggestion with 2 return values), then:

```python
class MyNet(NeuralNetRegressor):
    def get_loss(self, y_pred, y_true, **kwargs):
        y_pred = y_pred[0] if isinstance(y_pred, tuple) else y_pred  # discard the 2nd output
        return super().get_loss(y_pred, y_true, **kwargs)

net = MyNet(...)
net.fit(X, y)
y_pred, features = net.forward(X)
```

**Comment by junwoony** [comment]
Yes it works!
Thank you @BenjaminBossan and @ottonemo !!

**Comment by BenjaminBossan** [comment]
Should be fixed by #457, please re-open if there are still problems.
